lib/spack/docs/build_systems/gopackage.rst

```diff
@@ -0,0 +1,164 @@
+.. Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+   Spack Project Developers. See the top-level COPYRIGHT file for details.
+
+   SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+.. _gopackage:
+
+-----------
+GoPackage
+-----------
+
+There are a couple of things to be aware of about applications built
+with Go:
+
+* Go-based projects use a variety of build tools: simple invocations
+  of ``go build``, custom go "scripts" that leverage ``go run``, Make,
+  and *etc...*.
+
+* Go's dependency management story has evolved from an initial
+  do-it-yourself attitude through a small number of commonly used
+  tools to an officially supported approach (Go "modules").  The Go
+  team provides an `introduction to Go modules
+  <https://blog.golang.org/using-go-modules>`_ and maintains a site
+  that tracks `the gory details
+  <https://github.com/golang/go/wiki/Modules>`_.
+
+  Go modules are the future, ``GOPATH`` will likely be deprecated in
+  Go 1.14.
+
+Not surprisingly, the best approach to packaging a Go-based project
+for Spack depends on the details of the project.
+
+The ``GoPackage`` base class is intended to make it easy to package
+simple projects.  Tools & techniques that it uses may also be useful
+when using other build system base classes (e.g. ``MakefilePackage``),
+in particular:
+
+- Use the package's ``setup_build_environment`` to set
+  ``GO111MODULE=on`` to enable module mode and set
+  ``GOFLAGS="-mod-vendor"`` to enable vendoring, which prevents ``go``
+  from downloading libraries at build time.
+- If the application hasn't vendored its dependencies, use
+  ``modules2tuple`` to generate "bulk" resource statements and include
+  them in the package definition via ``import_resources``.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+Simple package definitions
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The simplest case is a project that uses a simple invocation of ``go
+build`` to compile an executable, *and* uses modules, *and* has
+vendored its dependencies.  Here's an example of a package definition
+which is only slightly complicated by the fact that it requires a
+newer version of Go than ``GoPackage`` provides by default:
+
+.. code-block:: python
+
+    from spack import *
+
+
+    class Lazygit(GoPackage):
+        """Simple terminal UI for git commands."""
+
+        homepage = "https://github.com/jesseduffield/lazygit"
+        url      = "https://github.com/jesseduffield/lazygit/archive/v0.11.3.tar.gz"
+
+        version('0.11.3', sha256='b3c503de6b34fd4284fd70655e7f42feafc07f090e7f7cc00db261f56c583c46')
+
+        depends_on('go@1.13:', type='build')  # go.mod value overrides default
+
+        executables = ['lazygit']
+
+While it generally follows the common patterns of a Spack package,
+it's worth calling attention to its:
+
+* adding a *build* dependency on the version of the Go package that it
+  requires (from the existing `go.mod`);
```

Comment:
In other build system docs, I added an entire section on "Important files", where you could describe what `go.mod` is, what to look for, and an example of how it specifies the minimum Go version needed. For example, see https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#important-files

Reply:
I'll improve this.

Reply:
Interestingly, waf.py, sip.py, and now go.py define and call it as:

```python
def build_args(self):
```

while scons.py and python.py define and call it as:

```python
def build_args(self, spec, prefix):
```

Reply:
done